ext/re2/re2.cc

```diff
@@ -122,18 +122,34 @@ static void parse_re2_options(RE2::Options* re2_options, const VALUE options) {
   }
 }
 
-static void re2_matchdata_mark(re2_matchdata* self) {
+static void re2_matchdata_mark(void * data) {
+  re2_matchdata* self = (re2_matchdata*)data;
   rb_gc_mark(self->regexp);
   rb_gc_mark(self->text);
 }
 
-static void re2_matchdata_free(re2_matchdata* self) {
+static void re2_matchdata_free(void * data) {
+  re2_matchdata* self = (re2_matchdata*)data;
   if (self->matches) {
     delete[] self->matches;
   }
   free(self);
 }
 
+static size_t re2_matchdata_memsize(const void * data) {
+  return sizeof(re2_matchdata);
```

Comment:
Should we be adding the size of the values in `data->matches` here too?

Reply:
Fixed in 8ded5101f913b5b29428916f7c57b6e3a6fa68b6.

Reply:
Actually, I think it should be `sizeof(re2::StringPiece) * self->number_of_matches`.

Not a huge deal though, the memsize is only for `ObjectSpace.memsize_of` and `ObjectSpace.dump`.

Reply:
I amended this in https://github.com/mudge/re2/commit/677352de905bd60f5616b77ffeeee8c4098eefa5 as I noticed the reported `memsize` from `ObjectSpace.dump` didn't change when there were more matches but that uses `sizeof(self->matches)` rather than `sizeof(re2::StringPiece)` as the basis of the calculation. I'll update it to use your suggestion as that seems slightly more accurate (though, as you say, this isn't a show-stopper).

Reply:
Updated (once again) in https://github.com/mudge/re2/commit/9cdb8a900d2fabd6d071055e3f1ecbd99698d114.

Reply:
@byroot does this mean we should also use `sizeof(*s->input)` rather than `sizeof(s->input)` to use the size of the `StringPiece` rather than the size of the _pointer_ to a `StringPiece` (same for `sizeof(p->pattern)` and `sizeof(s->set)` where they are both pointers)?

Reply:
Yes, indeed. (Sorry, I'm a terrible C developer, and even worse C++ dev).

Reply:
Basically, you should to account for all the structs that were `malloc` or otherwise allocated by that Object.